compose/config/fields_schema.json

```diff
@@ -37,7 +37,15 @@
 
         "environment": {
           "oneOf": [
-            {"type": "object"},
+            {
+              "type": "object",
+              "patternProperties": {
+                "^[a-zA-Z0-9_]+$": {
```

Comment:
I think we should actually relax this to be any character (or maybe anything but dash).  While bash may not support everything, swarm at least needs colons (http://docs.docker.com/swarm/scheduler/filter/#affinity-filter), so it might be easier to just allow most characters for now.